### PR TITLE
Don't allow a second close of a PipeConn

### PR DESCRIPTION
**Pull request by maxtaco**
- Also some tweaks to get tests working on my box.

### CODE REVIEW

File: npipe_windows.go [hunk 1]
```diff
@@ -443,6 +446,12 @@ func (c *PipeConn) Write(b []byte) (int, error) {
 
 // Close closes the connection.
 func (c *PipeConn) Close() error {
+	c.closedMutex.Lock()
+	defer c.closedMutex.Unlock()
+	if c.closed {
+		return nil
+	}
+	c.closed = true
 	return syscall.CloseHandle(c.handle)
```
Comment (zanderz) [inline npipe_windows.go:455]: Strictly speaking, we're unnecessarily holding the mutex while calling out to CloseHandle. An atomic compare/exchange would be better.

Comment (maxtaco) [inline npipe_windows.go:455]: That's true. I don't think Go gives you atomic CAS, I'll just release the lock sooner, but I didn't htink it mattered.

**Comment by zanderz** [comment]
wow, now I remember catching a multiple close here. Can't remember why I didn't do more with that.

LGTM :thumbsup:

File: npipe_windows.go (hunk 1 above)
Comment (maxtaco) [inline npipe_windows.go:455]: Well, it does give you CAS, but tells you not to bother with it :).  I'll move those locks around.

Comment (zanderz) [inline npipe_windows.go:455]: I don't think it matters really, just making the observation. Too bad you can't do an early defer or something in Go.

Comment (zanderz) [inline npipe_windows.go:455]: I took another look at the atomic library too - no compare and exchange for bool :(

Comment (maxtaco) [inline npipe_windows.go:455]: Makes the code alot uglier to not do it this way.  I can't imagine any deadlock by having the second guy wait for the first to complete the system call.

**Pull request merged**